src/team_lib/hier/xccl_hier_lib.c

```diff
@@ -15,31 +15,44 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+static ucs_config_field_t xccl_tl_hier_context_config_table[] = {
+    {"", "", NULL,
+        ucs_offsetof(xccl_tl_hier_context_config_t, super),
+        UCS_CONFIG_TYPE_TABLE(xccl_tl_context_config_table)
+    },
+
+    {"NET_DEVICES", "all",
+     "Specifies which network device(s) to use",
+     ucs_offsetof(xccl_tl_hier_context_config_t, devices), UCS_CONFIG_TYPE_STRING_ARRAY
+    },
+
+    {NULL}
+};
 
 static inline xccl_status_t
 xccl_hier_allreduce_init(xccl_coll_op_args_t *coll_args,
                          xccl_coll_req_h *request, xccl_tl_team_t *team)
 {
     //TODO alg selection for allreduce shoud happen here
     coll_schedule_t *schedule;
-    xccl_hier_context_t *ctx = xccl_derived_of(team->ctx, xccl_hier_context_t);
+    xccl_hier_context_t *ctx = ucs_derived_of(team->ctx, xccl_hier_context_t);
     xccl_hier_allreduce_spec_t spec = {
         .pairs              = {
-            .node_leaders   = ctx->tls[XCCL_TL_SHARP].enabled ?
+            .node_leaders   = ctx->tls[ucs_ilog2(XCCL_TL_SHARP)].enabled ?
                               XCCL_HIER_PAIR_NODE_LEADERS_SHARP :
                               XCCL_HIER_PAIR_NODE_LEADERS_UCX,
-            .socket         = ctx->tls[XCCL_TL_SHMSEG].enabled ?
+            .socket         = ctx->tls[ucs_ilog2(XCCL_TL_SHMSEG)].enabled ?
                               XCCL_HIER_PAIR_SOCKET_SHMSEG :
                               XCCL_HIER_PAIR_SOCKET_UCX,
-            .socket_leaders = ctx->tls[XCCL_TL_SHMSEG].enabled ?
+            .socket_leaders = ctx->tls[ucs_ilog2(XCCL_TL_SHMSEG)].enabled ?
                               XCCL_HIER_PAIR_SOCKET_LEADERS_SHMSEG :
                               XCCL_HIER_PAIR_SOCKET_LEADERS_UCX,
         },
     };
-    build_allreduce_schedule(xccl_derived_of(team, xccl_hier_team_t), (*coll_args),
+    build_allreduce_schedule(ucs_derived_of(team, xccl_hier_team_t), (*coll_args),
                              spec, &schedule);
     schedule->super.lib = &xccl_team_lib_hier.super;
-    (*request) = &schedule->super;
+    (*request) = (xccl_coll_req_h)&schedule->super;
```

Comment:
maybe instead of cast use &schedule->super.req
btw, why in tl_coll_req the super is called "req" (and not super again) ?

Reply:
You mean xccl_coll_req? We need to discuss it but the idea was to define xccl_coll_req as we define xccl_team or xccl_context as collection of particular team libs object and encapsulate team_lib request inside xccl request. So xccl_coll_req is not derived from team_lib req but just stores it.